### DISCUSSION TITLE
What is the ultimate "Source" of Application kind itself?

### DISCUSSION THREAD

**Discussion by Veluu** [argoproj/argo-cd#11147] [Q&A]
Hello

I want to understand what ArgoCD is syncing application to in the following scenario. By application I mean `kind: Application` itself.

Application is defined as follows and deployed by Terraform

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: aws-node-termination-handler
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: management
  source:
    repoURL: "https://aws.github.io/eks-charts"
    targetRevision: v0.18.4
    chart: aws-node-termination-handler
  destination:
    server: https://kubernetes.default.svc
    namespace: kube-system
  syncPolicy:
    automated:
      prune: true # ( false by default ).
      selfHeal: true # ( false by default ).
      # allowEmpty: false #  ( false by default ).
    syncOptions:
      - CreateNamespace=false
```
I do apply it with Terraform and https://github.com/gavinbunney/terraform-provider-kubectl. Everything is fine. I do re-run `terraform apply` and manual `Sync` from ArgoCD UI. Still everything is fine.

I'm checking my application definition in Kubernetes (I ommit `status` in Application for readability).

```bash
 get -n argocd Application aws-node-termination-handler -oyaml
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1alpha1","kind":"Application","metadata":{"annotations":{},"finalizers":["resources-finalizer.argocd.argoproj.io"],"name":"aws-node-termination-handler","namespace":"argocd"},"spec":{"destination":{"namespace":"kube-system","server":"https://kubernetes.default.svc"},"project":"management","source":{"chart":"aws-node-termination-handler","repoURL":"https://aws.github.io/eks-charts","targetRevision":"v0.18.4"},"syncPolicy":{"automated":{"prune":true,"selfHeal":true},"syncOptions":["CreateNamespace=false"]}}}
  creationTimestamp: "2022-11-01T21:47:55Z"
  finalizers:
  - resources-finalizer.argocd.argoproj.io
  generation: 15
  name: aws-node-termination-handler
  namespace: argocd
  resourceVersion: "1768924"
  uid: 58aaf7c2-5f60-46d5-9e7e-e199c9e7c9b9
spec:
  destination:
    namespace: kube-system
    server: https://kubernetes.default.svc
  project: management
  source:
    chart: aws-node-termination-handler
    repoURL: https://aws.github.io/eks-charts
    targetRevision: v0.18.4
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
    - CreateNamespace=false

```

Now, I uncomment `allowEmpty: false` in the Application definition and run `terraform apply`. Update is made. I check what's in Kuberentes.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1alpha1","kind":"Application","metadata":{"annotations":{},"finalizers":["resources-finalizer.argocd.argoproj.io"],"name":"aws-node-termination-handler","namespace":"argocd"},"spec":{"destination":{"namespace":"kube-system","server":"https://kubernetes.default.svc"},"project":"management","source":{"chart":"aws-node-termination-handler","repoURL":"https://aws.github.io/eks-charts","targetRevision":"v0.18.4"},"syncPolicy":{"automated":{"allowEmpty":false,"prune":true,"selfHeal":true},"syncOptions":["CreateNamespace=false"]}}}
  creationTimestamp: "2022-11-01T21:47:55Z"
  finalizers:
  - resources-finalizer.argocd.argoproj.io
  generation: 28
  name: aws-node-termination-handler
  namespace: argocd
  resourceVersion: "1769570"
  uid: 58aaf7c2-5f60-46d5-9e7e-e199c9e7c9b9
spec:
  destination:
    namespace: kube-system
    server: https://kubernetes.default.svc
  project: management
  source:
    chart: aws-node-termination-handler
    repoURL: https://aws.github.io/eks-charts
    targetRevision: v0.18.4
  syncPolicy:
    automated:
      allowEmpty: false
      prune: true
      selfHeal: true
    syncOptions:
    - CreateNamespace=false
```

Looks right. I do go to ArgoCD web ui. I click Sync on the app. 

```bash
 get -n argocd Application aws-node-termination-handler -oyaml
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1alpha1","kind":"Application","metadata":{"annotations":{},"finalizers":["resources-finalizer.argocd.argoproj.io"],"name":"aws-node-termination-handler","namespace":"argocd"},"spec":{"destination":{"namespace":"kube-system","server":"https://kubernetes.default.svc"},"project":"management","source":{"chart":"aws-node-termination-handler","repoURL":"https://aws.github.io/eks-charts","targetRevision":"v0.18.4"},"syncPolicy":{"automated":{"allowEmpty":false,"prune":true,"selfHeal":true},"syncOptions":["CreateNamespace=false"]}}}
  creationTimestamp: "2022-11-01T21:47:55Z"
  finalizers:
  - resources-finalizer.argocd.argoproj.io
  generation: 34
  name: aws-node-termination-handler
  namespace: argocd
  resourceVersion: "1769942"
  uid: 58aaf7c2-5f60-46d5-9e7e-e199c9e7c9b9
spec:
  destination:
    namespace: kube-system
    server: https://kubernetes.default.svc
  project: management
  source:
    chart: aws-node-termination-handler
    repoURL: https://aws.github.io/eks-charts
    targetRevision: v0.18.4
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
    - CreateNamespace=false
```
Seems like ArgoCD is deleting `syncPolicy.automated.allowEmpty: false` value.

I re-run Terraform and it wants to make an update based on the yaml file I loaded in Terraform - it wants to re-add `syncPolicy.automated.allowEmpty: false`. I don't blame it - it's what in the definition.

Long story short:
- Initially I had only provided "prune":true,"selfHeal":true
- I updated that to become   "allowEmpty":false, "prune":true,"selfHeal":true
- That created a desynch of what ArgoCD sees vs what Terraform sees.

I understand that when I modify files that `Application` is pointing to ArgoCD will sync to those changes. In my example I'm modifying the `Application` kind itself.

I also understand that most likely if I followed the app of app pattern I would have not faced this issue. Nevertheless, could somebody explain to me what ArgoCD is syncing kind `Application` to? As it seems that once created I should not make any changes to `Application` configuration as I will run into discrepancy.
